# Random Forest Classification

Using the forest of decision trees in getting the species of penguins. The data set used is gotting from seaborn library "Penguis" data. Using the three different island  which are Biscoe,	Dream,	Torgersen. The species of the penguins bird was discovcered by the decision nodes which are body mass, and flipper length. These decision serve as the basis for the classification of the species of the bird. 

In [5]:
import pandas as pd
import numpy as np
import seaborn as sns 

### Load Data Set from seaborn library

In [6]:
df = sns.load_dataset('penguins')

df.head(10)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   
6  Adelie  Torgersen            38.9           17.8              181.0   
7  Adelie  Torgersen            39.2           19.6              195.0   
8  Adelie  Torgersen            34.1           18.1              193.0   
9  Adelie  Torgersen            42.0           20.2              190.0   

   body_mass_g     sex  
0       3750.0    Male  
1       3800.0  Female  
2       3250.0  Female  
3          NaN     NaN  
4       3450.0  Female  
5       3650.0    Male  
6       3625.0  Female  
7       4675.0    Male  
8       3475.0     NaN  
9       4250.0     NaN

In [7]:
df.shape

(344, 7)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [9]:
df.dropna(inplace=True)

In [10]:
df.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
dtype: int64

## Feature Engineering

### Transform the categorical data into numeric using One Hot

In [11]:
df['sex'].unique()

array(['Male', 'Female'], dtype=object)

In [12]:
pd.get_dummies(df.sex).head(4)

Female  Male
0       0     1
1       1     0
2       1     0
4       1     0

In [13]:
sex = pd.get_dummies(df['sex'], drop_first =True)
sex.head()

Male
0     1
1     0
2     0
4     0
5     1

### Apply one hot encoding to island

In [14]:
df['island'].unique()

array(['Torgersen', 'Biscoe', 'Dream'], dtype=object)

In [15]:
pd.get_dummies(df['island']).head()

Biscoe  Dream  Torgersen
0       0      0          1
1       0      0          1
2       0      0          1
4       0      0          1
5       0      0          1

In [16]:
island = pd.get_dummies(df['island'], drop_first = True)
island.head(4)

Dream  Torgersen
0      0          1
1      0          1
2      0          1
4      0          1

### Concatenate the date frame into original data

In [17]:
new_data = pd.concat([df,island,sex],axis =1)

In [18]:
new_data.head(5)

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
4  Adelie  Torgersen            36.7           19.3              193.0   
5  Adelie  Torgersen            39.3           20.6              190.0   

   body_mass_g     sex  Dream  Torgersen  Male  
0       3750.0    Male      0          1     1  
1       3800.0  Female      0          1     0  
2       3250.0  Female      0          1     0  
4       3450.0  Female      0          1     0  
5       3650.0    Male      0          1     1

### Drop the repeated column

In [19]:
new_data.drop(['island','sex'], axis= 1, inplace =True)

In [20]:
new_data.head()

species  bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
0  Adelie            39.1           18.7              181.0       3750.0   
1  Adelie            39.5           17.4              186.0       3800.0   
2  Adelie            40.3           18.0              195.0       3250.0   
4  Adelie            36.7           19.3              193.0       3450.0   
5  Adelie            39.3           20.6              190.0       3650.0   

   Dream  Torgersen  Male  
0      0          1     1  
1      0          1     0  
2      0          1     0  
4      0          1     0  
5      0          1     1

### Creating Separate target variable

In [21]:
Y= new_data['species']
Y.head()

0    Adelie
1    Adelie
2    Adelie
4    Adelie
5    Adelie
Name: species, dtype: object

In [22]:
Y.unique()

array(['Adelie', 'Chinstrap', 'Gentoo'], dtype=object)

### Convert the target variable to numeric using map function


In [23]:
Y = Y.map({'Adelie':0, 'Chinstrap':1, 'Gentoo':2})
Y.head(5)

0    0
1    0
2    0
4    0
5    0
Name: species, dtype: int64

### Droping the target variable for our data frame

In [24]:
new_data.drop('species', axis =1, inplace=True)

In [25]:
new_data.head(5)

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  Dream  \
0            39.1           18.7              181.0       3750.0      0   
1            39.5           17.4              186.0       3800.0      0   
2            40.3           18.0              195.0       3250.0      0   
4            36.7           19.3              193.0       3450.0      0   
5            39.3           20.6              190.0       3650.0      0   

   Torgersen  Male  
0          1     1  
1          1     0  
2          1     0  
4          1     0  
5          1     1

In [26]:
X =  new_data

### Spliting the data into training and test data

In [27]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.3, random_state=0)

In [28]:
print('X_train', X_train.shape)
print('X_test', X_test.shape)
print('Y_train', Y_train.shape)
print('Y_test', Y_test.shape)

X_train (233, 7)
X_test (100, 7)
Y_train (233,)
Y_test (100,)


### Training Random forest classification on Training set

In [29]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 5, criterion = 'entropy', random_state = 0) ### 5 is 5 decision tree
classifier.fit(X_train, Y_train)

RandomForestClassifier(criterion='entropy', n_estimators=5, random_state=0)

### Making some prediction on test results 

In [30]:
y_pred =classifier.predict(X_test)

y_pred

array([0, 0, 2, 0, 0, 0, 1, 2, 2, 1, 2, 0, 0, 1, 0, 0, 2, 0, 1, 0, 0, 0,
       2, 2, 2, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 1, 0, 2, 2, 0, 0,
       0, 0, 0, 0, 2, 0, 0, 0, 2, 2, 0, 0, 0, 0, 0, 2, 0, 1, 0, 2, 0, 0,
       2, 2, 1, 2, 2, 1, 2, 1, 0, 2, 0, 2, 0, 2, 1, 2, 2, 2, 1, 2, 1, 0,
       0, 2, 2, 0, 2, 0, 2, 0, 2, 0, 2, 2], dtype=int64)

### Confusion matrix (to check the accuracy of random forest algorithm)

In [31]:
from sklearn.metrics import confusion_matrix 
from sklearn.metrics import classification_report, accuracy_score

In [32]:
cm = confusion_matrix(Y_test, y_pred)
print(cm)

[[48  0  0]
 [ 2 14  0]
 [ 0  0 36]]


In [33]:
accuracy_score(Y_test, y_pred)

0.98

In [34]:
print(classification_report(Y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98        48
           1       1.00      0.88      0.93        16
           2       1.00      1.00      1.00        36

    accuracy                           0.98       100
   macro avg       0.99      0.96      0.97       100
weighted avg       0.98      0.98      0.98       100



### Let try different number of decision tree and change the criteria to 

In [35]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 7, criterion = 'gini', random_state = 0) ### 7 is 7 decision tree
classifier.fit(X_train, Y_train)

RandomForestClassifier(n_estimators=7, random_state=0)

In [36]:
y_pred = classifier.predict(X_test)

In [37]:
accuracy_score(Y_test, y_pred)

0.99

In [38]:
print(classification_report(Y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      1.00      0.99        48
           1       1.00      0.94      0.97        16
           2       1.00      1.00      1.00        36

    accuracy                           0.99       100
   macro avg       0.99      0.98      0.99       100
weighted avg       0.99      0.99      0.99       100



Increasing the number of decions tree we got a higher test accuracy of the trianing data set. Also, a better precision of the training data. Using 5 and 7 decision tree in the both test it giving a test accuracy of 96% and 99% respectively. Which show our random selection of the decions tree gives the actual species of the penguins select from diffrent island.  